sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/SplitGenerator.java

```diff
@@ -0,0 +1,179 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(SplitGenerator.class);
+
+  private final String partitioner;
+  private final BigInteger rangeMin;
+  private final BigInteger rangeMax;
+  private final BigInteger rangeSize;
+
+  SplitGenerator(String partitioner){
+    rangeMin = getRangeMin(partitioner);
+    rangeMax = getRangeMax(partitioner);
+    rangeSize = getRangeSize(partitioner);
+    this.partitioner = partitioner;
+  }
+
+  SplitGenerator(BigInteger rangeMin, BigInteger rangeMax) {
+    this.rangeMin = rangeMin;
+    this.rangeMax = rangeMax;
+    rangeSize = rangeMax.subtract(rangeMin).add(BigInteger.ONE);
+    partitioner = "(" + rangeMin + "," + rangeMax + ")";
+  }
+
+  static BigInteger getRangeMin(String partitioner) {
+    if (partitioner.endsWith("RandomPartitioner")) {
+      return BigInteger.ZERO;
+    } else if (partitioner.endsWith("Murmur3Partitioner")) {
+      return new BigInteger("2").pow(63).negate();
+    } else {
+      throw new UnsupportedOperationException("Unsupported partitioner. "
+          + "Only Random and Murmur3 are supported");
+    }
+  }
+
+  static BigInteger getRangeMax(String partitioner) {
+    if (partitioner.endsWith("RandomPartitioner")) {
+      return new BigInteger("2").pow(127).subtract(BigInteger.ONE);
+    } else if (partitioner.endsWith("Murmur3Partitioner")) {
+      return new BigInteger("2").pow(63).subtract(BigInteger.ONE);
+    } else {
+      throw new UnsupportedOperationException("Unsupported partitioner. "
+          + "Only Random and Murmur3 are supported");
+    }
+  }
+
+  static BigInteger getRangeSize(String partitioner) {
+    return getRangeMax(partitioner).subtract(getRangeMin(partitioner)).add(BigInteger.ONE);
+  }
+
+  static BigInteger max(BigInteger big0, BigInteger big1) {
+    return greaterThan(big0, big1) ? big0 : big1;
+  }
+
+  static BigInteger min(BigInteger big0, BigInteger big1) {
+    return lowerThan(big0, big1) ? big0 : big1;
+  }
+
+  static boolean lowerThan(BigInteger big0, BigInteger big1) {
+    return big0.compareTo(big1) < 0;
+  }
+
+  static boolean lowerThanOrEqual(BigInteger big0, BigInteger big1) {
+    return big0.compareTo(big1) <= 0;
+  }
+
+  static boolean greaterThan(BigInteger big0, BigInteger big1) {
+    return big0.compareTo(big1) > 0;
+  }
+
+  static boolean greaterThanOrEqual(BigInteger big0, BigInteger big1) {
+    return big0.compareTo(big1) >= 0;
+  }
+
+  /**
+   * Given big0 properly ordered list of tokens, compute at least {@code totalSplitCount} splits.
+   *
+   * @param totalSplitCount requested total amount of splits. This function may generate
+   *                        more splits.
+   * @param ringTokens list of all start tokens in big0 cluster. They have to be in ring order.
+   * @return big0 list containing at least {@code totalSplitCount} splits.
+   */
+  List<RingRange> generateSplits(long totalSplitCount, List<BigInteger> ringTokens) {
+
+    int tokenRangeCount = ringTokens.size();
```

Comment:
I found an inconsistency in the split production to reproduce it run the CassandraIOIT tests (you can hardcore the options to run it quickly in IntelliJ in the `@Setup`..

This is how to reproduce
`CassandraTestDataSet` produces the table with data (1000 records) then run `CassandraIOIT#testRead` but pass the `.withMinNumberOfSplits(512)` option. It will run for some time and at the end it will find 998 records. Not sure what is going on.

Reply:
@jkff since you are quite good to find tricky corner cases would you mind to take a look just at this method (I will take care of the rest of the PR).

Reply:
@iemejia : I've just pushed the fix.
I hadn't handled the wrapping range correctly (the one that covers both the end and the start of the token ring).
That range now generates two valid queries : one for all tokens greater than or equal to the start token, and another for all tokens lower than the end token.

The IT are passing now.